File: Java/Singly_Linked_List/SLL.java
public class SLL {
    private Node start;

    public SLL() {
        this.start = null;
    }

    public void add(Node node) {
        if (this.start == null) {
            this.start = node;
            return;
        }

        Node pointer = this.start;

        while (pointer.hasNext()) {
            pointer = pointer.getNext();
        }

        pointer.setNext(node);
    }

    public void remove(Node node) {
        if (this.start == null) {
            return;
        }

        if (!this.start.hasNext() && this.start == node) {
            this.start = null;
            return;
        }

        Node prev = null;
        Node cur = this.start;

        while (cur != null && !cur.equals(node)) {
            prev = cur;
            cur = cur.getNext();
        }

        if (cur == this.start) {
            this.start = this.start.getNext();
            return;
        }

        if (cur == null) {
            return;
        }

        prev.setNext(cur.getNext());
        cur.setNext(null);
    }

    public void printSLL() {
        if (this.start == null) {
            return;
        }

        Node pointer = this.start;
        System.out.print("[ ");

        while (pointer != null) {
            System.out.print(pointer.getVal() + " ");
            pointer = pointer.getNext();
        }

        System.out.println("]");
    }
}

File: Python/sll.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class SLL:
    def __init__(self):
        self.head = None

    def add(self, val):
        node = Node(val)

        if not self.head:
            self.head = node
            return

        cur = self.head
        while cur.next:
            cur = cur.next

        cur.next = node

    def delete(self, val):
        if not self.head:
            return

        prev = None
        cur = self.head

        while cur and cur.val != val:
            prev = cur
            cur = cur.next

        if cur == self.head:
            self.head = self.head.next
            return

        if not cur:
            return

        prev.next = cur.next
        del cur

    def __str__(self):
        cur = self.head
        string = "[ "

        while cur:
            string += str(cur.val) + " "
            cur = cur.next

        string += "]"

        return string


range_ = 4
sll = SLL()

for i in range(range_):
    sll.add(i)

print(sll)

# sll.delete(3)
# print(sll)

# sll.delete(0)
# print(sll)

for i in range(range_):
    sll.delete(i)

print(sll)
